### DISCUSSION TITLE
MiniRocket Data format questions

### DISCUSSION THREAD

**Discussion by YojoNick** [timeseriesAI/tsai#165] [Q&A] [answered]
In the [MiniRocket tutorial notebook](https://github.com/timeseriesAI/tsai/blob/main/tutorial_nbs/10_Time_Series_Classification_and_Regression_with_MiniRocket.ipynb), is the variable X in the following code formatted as follows?

Here's the Code:
```
mrf = MiniRocketFeatures(X.shape[1], X.shape[2]).to(default_device())
X_train = X[splits[0]]
mrf.fit(X_train)
X_feat = get_minirocket_features(X, mrf, chunksize=1024, to_np=True)
```

Is this how the variable X is formatted?
Number_of_Samples x Dimensions x Sequence_Length

So, if I had 10 training samples, that each had 11 time steps with 2 dimensions per time step, then would X's shape be 10x2x11?

Is it assumed that X is ordered such that the time sequence is ascending?  i.e. the first elements along the third dimension sequence_length are the oldest in time?  i.e. X[0,:,0] would be the oldest time step for the first sample and X[0,:,-1] would be the most recent time step?

Do these time steps have to occur at even intervals or can they occur at uneven intervals?  Meaning the elements along the sequence_length dimension, do they have to exist at say seconds 1, 2, 3, 4, etc.  Or can they exist at arbitrary time steps - seconds 1.2, 1.5, 2.7, 4.5, etc.

**Comment by oguiza** [comment]
Hi @YojoNick,

I'l try to answer all your questions:

> Is this how the variable X is formatted?
   Number_of_Samples x Dimensions x Sequence_Length

Yes

> So, if I had 10 training samples, that each had 11 time steps with 2 dimensions per time step, then would X's shape be 10x2x11?

Yes

> Is it assumed that X is ordered such that the time sequence is ascending? i.e. the first elements along the third dimension sequence_length are the oldest in time? i.e. X[0,:,0] would be the oldest time step for the first sample and X[0,:,-1] would be the most recent time step?

It doesn't really matter whether it's ascending or descending as long as you consistently apply the same approach (during training and inference). Usually the first element is the oldest one though.


> Do these time steps have to occur at even intervals or can they occur at uneven intervals? Meaning the elements along the sequence_length dimension, do they have to exist at say seconds 1, 2, 3, 4, etc. Or can they exist at arbitrary time steps - seconds 1.2, 1.5, 2.7, 4.5, etc.

This is very interesting. I'd say it depends on the model. For CNNs and MiniRocket (which applies a 10_000 non-learnable convolutions) I'd say they need to be evenly distributed. 
However, with a Transformer, they could be unevenly distributed (even unsorted) as long as you add one variable that captures the time. In that case, the model should be able to learn to embed the time.

**Comment by YojoNick** [comment]
@oguiza that is super helpful.  Thank you!

**Comment by YojoNick** [comment, marked as answer]
Btw, the above definitely answered my questions.  I'm kind of a new to github discussions.  I don't know if I mark it as answered or someone else does.  Regardless, thanks again for the prompt and in depth response!

**Reply by oguiza** [reply]
Discussions have been recently added to GitHub AFAIK. I think anyone can flag it as as answer. I've done it, but feel free to do it yourself if you feel it's been answered. Actually, only the one who asks the question knows if it's truly been answered.

**Comment by YojoNick** [comment]
For anyone who stumbles across this thread, the format of the data is documented on this tutorial notebook:
https://colab.research.google.com/github/timeseriesAI/tsai/blob/master/tutorial_nbs/01_Intro_to_Time_Series_Classification.ipynb
